File: mac/Sources/LanClipCore/Snapshot.swift
import Foundation

public enum SnapshotError: Error, Equatable {
    case staleSeq
}

/// Итог `SnapshotStore.blob(index:seq:)` — находка I6 финального ревью: раньше
/// метод всегда возвращал `Data?`, читая файлы блобов `files` целиком в память
/// (`Data(contentsOf:)`) внутри `blob(...)` независимо от вызывающей стороны.
/// `maxBytes` — проверка на стороне КЛИЕНТА, поэтому `GET /clip/blob/{i}` можно
/// послать напрямую, в обход неё; многогигабайтное видео в буфере пользователя
/// вырубило бы сервер одним таким запросом, ещё до попытки отдать хоть байт.
/// `.file` заставляет вызывающую сторону (`HttpServer`) читать файл потоком, не
/// требуя от `SnapshotStore` знать что-либо о сети.
public enum BlobPayload: Sendable {
    case data(Data)
    case file(url: URL, size: Int)
}

public struct ClipSnapshot: Sendable {
    public let manifest: Manifest
    public let text: String?
    public let imagePNG: Data?
    public let sources: [Int: URL]
}

public final class SnapshotStore {
    private let reader: ClipboardReading
    private let fileManager: FileManager
    private var cached: ClipSnapshot?

    public init(reader: ClipboardReading, fileManager: FileManager = .default) {
        self.reader = reader
        self.fileManager = fileManager
    }

    public func current() throws -> ClipSnapshot {
        let seq = reader.changeCount()
        if let cached, cached.manifest.seq == seq { return cached }

        let snapshot = try build(seq: seq, content: try reader.read())
        cached = snapshot
        return snapshot
    }

    public func blob(index: Int, seq: Int) throws -> BlobPayload? {
        let snapshot = try current()
        guard snapshot.manifest.seq == seq else { throw SnapshotError.staleSeq }

        if let png = snapshot.imagePNG {
            return index == 0 ? .data(png) : nil
        }
        guard let source = snapshot.sources[index] else { return nil }
        // Размер считается заново с диска (не из манифеста) — это тот же файл,
        // который сейчас будет прочитан, поэтому `Content-Length`, ушедший в
        // заголовках, гарантированно совпадает с тем, что реально станет читать
        // `HttpServer` дальше чанками.
        return .file(url: source, size: try size(of: source))
    }

    private func build(seq: Int, content: ClipContent) throws -> ClipSnapshot {
        switch content {
        case .empty:
            return ClipSnapshot(manifest: .empty(seq: seq), text: nil, imagePNG: nil, sources: [:])

        case .text(let value):
            return ClipSnapshot(manifest: .text(value, seq: seq), text: value,
                                imagePNG: nil, sources: [:])

        case .image(let png):
            return ClipSnapshot(manifest: .image(pngSize: png.count, seq: seq), text: nil,
                                imagePNG: png, sources: [:])

        case .files(let urls):
            var blobs: [BlobRef] = []
            var sources: [Int: URL] = [:]

            for url in urls {
                for entry in try expand(url) {
                    guard let rel = RelPath.normalize(entry.rel) else { continue }
                    sources[blobs.count] = entry.url
                    blobs.append(BlobRef(i: blobs.count, rel: rel, size: entry.size))
                }
            }

            if blobs.isEmpty {
                return ClipSnapshot(manifest: .empty(seq: seq), text: nil, imagePNG: nil, sources: [:])
            }
            return ClipSnapshot(manifest: .files(blobs, seq: seq), text: nil,
                                imagePNG: nil, sources: sources)
        }
    }

    private struct Entry {
        let url: URL
        let rel: String
        let size: Int
    }

    private func expand(_ url: URL) throws -> [Entry] {
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: url.path, isDirectory: &isDirectory) else { return [] }

        if !isDirectory.boolValue {
            let resolved = url.resolvingSymlinksInPath()
            return [Entry(url: resolved, rel: url.lastPathComponent, size: try size(of: resolved))]
        }

        let base = url.lastPathComponent
        guard let walker = fileManager.enumerator(atPath: url.path) else {
            return []
        }

        var entries: [Entry] = []
        for case let subpath as String in walker {
            let child = url.appendingPathComponent(subpath)
            var isChildDirectory: ObjCBool = false
            guard fileManager.fileExists(atPath: child.path, isDirectory: &isChildDirectory) else { continue }
            if isChildDirectory.boolValue { continue }
            let resolved = child.resolvingSymlinksInPath()
            entries.append(Entry(url: resolved, rel: base + "/" + subpath, size: try size(of: resolved)))
        }
        return entries
    }

    private func size(of url: URL) throws -> Int {
        let attributes = try fileManager.attributesOfItem(atPath: url.path)
        return (attributes[.size] as? NSNumber)?.intValue ?? 0
    }
}

File: win/src/Snapshot.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace LanClip
{
    // Зеркало mac/Sources/LanClipCore/Snapshot.swift (задача 5). Два дефекта,
    // найденных ревью на Swift-стороне, сюда не переносятся:
    //
    //  1. Относительные пути строятся из позиции обхода (relPrefix + "/" + имя),
    //     а не вычитанием префикса из полного пути — так temp-каталог, зарезолвленный
    //     через junction/подставленный диск, не даёт мусора в rel.
    //  2. Размер и отдаваемые байты берутся из одной и той же строки пути: Expand()
    //     вычисляет её один раз (CanonicalizeForBothReads — чисто лексическая
    //     нормализация, не разыменование symlink/junction) и кладёт эту же строку и
    //     в BlobRef.Size (через FileInfo), и в Sources (для последующего чтения в
    //     Blob()). Расхождение между "что посчитали" и "что прочитали" структурно
    //     невозможно — это буквально один и тот же путь, а не два разных API с
    //     разным поведением на symlink, как было в исходном Swift-баге.

    class ClipSnapshot
    {
        public Manifest Manifest;
        public byte[] ImagePng;
        public Dictionary<int, string> Sources;
    }

    class StaleSeqException : Exception
    {
        public StaleSeqException()
            : base("устаревший seq: буфер уже изменился, нужен свежий снимок")
        {
        }
    }

    class SnapshotStore
    {
        readonly IClipboardReader reader;
        ClipSnapshot cached;

        public SnapshotStore(IClipboardReader reader)
        {
            this.reader = reader;
        }

        public ClipSnapshot Current()
        {
            int seq = reader.ChangeCount();
            if (cached != null && cached.Manifest.Seq == seq)
            {
                return cached;
            }

            ClipSnapshot snapshot = Build(seq, reader.Read());
            cached = snapshot;
            return snapshot;
        }

        // null — индекс вне диапазона (нет такого blob-а в текущем снимке).
        public byte[] Blob(int index, int seq)
        {
            ClipSnapshot snapshot = Current();
            if (snapshot.Manifest.Seq != seq)
            {
                throw new StaleSeqException();
            }

            if (snapshot.ImagePng != null)
            {
                return index == 0 ? snapshot.ImagePng : null;
            }

            string source;
            if (!snapshot.Sources.TryGetValue(index, out source))
            {
                return null;
            }
            return File.ReadAllBytes(source);
        }

        static ClipSnapshot Build(int seq, ClipContent content)
        {
            if (content.Kind == ClipKindValue.Empty)
            {
                return EmptySnapshot(seq);
            }

            if (content.Kind == ClipKindValue.Text)
            {
                ClipSnapshot snapshot = new ClipSnapshot();
                snapshot.Manifest = Manifest.OfText(content.Text, seq);
                snapshot.Sources = new Dictionary<int, string>();
                return snapshot;
            }

            if (content.Kind == ClipKindValue.Image)
            {
                ClipSnapshot snapshot = new ClipSnapshot();
                snapshot.Manifest = Manifest.OfImage(content.Png.Length, seq);
                snapshot.ImagePng = content.Png;
                snapshot.Sources = new Dictionary<int, string>();
                return snapshot;
            }

            return BuildFiles(seq, content.Files);
        }

        static ClipSnapshot EmptySnapshot(int seq)
        {
            ClipSnapshot snapshot = new ClipSnapshot();
            snapshot.Manifest = Manifest.Empty(seq);
            snapshot.Sources = new Dictionary<int, string>();
            return snapshot;
        }

        static ClipSnapshot BuildFiles(int seq, List<string> paths)
        {
            List<BlobRef> blobs = new List<BlobRef>();
            Dictionary<int, string> sources = new Dictionary<int, string>();

            foreach (string path in paths)
            {
                foreach (Entry entry in Expand(path))
                {
                    string rel = RelPath.Normalize(entry.Rel);
                    if (rel == null)
                    {
                        continue;
                    }

                    BlobRef blob = new BlobRef();
                    blob.I = blobs.Count;
                    blob.Rel = rel;
                    blob.Size = entry.Size;

                    sources[blobs.Count] = entry.AbsolutePath;
                    blobs.Add(blob);
                }
            }

            if (blobs.Count == 0)
            {
                return EmptySnapshot(seq);
            }

            ClipSnapshot snapshot = new ClipSnapshot();
            snapshot.Manifest = Manifest.OfFiles(blobs, seq);
            snapshot.Sources = sources;
            return snapshot;
        }

        struct Entry
        {
            public string AbsolutePath;
            public string Rel;
            public long Size;
        }

        static List<Entry> Expand(string path)
        {
            List<Entry> entries = new List<Entry>();

            if (File.Exists(path))
            {
                string absolute = CanonicalizeForBothReads(path);
                Entry entry = new Entry();
                entry.AbsolutePath = absolute;
                entry.Rel = Path.GetFileName(path);
                entry.Size = new FileInfo(absolute).Length;
                entries.Add(entry);
                return entries;
            }

            if (!Directory.Exists(path))
            {
                return entries;
            }

            string baseName = new DirectoryInfo(path).Name;
            Walk(path, baseName, entries);
            return entries;
        }

        // Обходит дерево рекурсивно, строя rel из позиции в обходе (relPrefix +
        // "/" + имя), а не арифметикой над строками полного пути.
        static void Walk(string dir, string relPrefix, List<Entry> entries)
        {
            foreach (string filePath in Directory.GetFiles(dir))
            {
                string absolute = CanonicalizeForBothReads(filePath);
                Entry entry = new Entry();
                entry.AbsolutePath = absolute;
                entry.Rel = relPrefix + "/" + Path.GetFileName(filePath);
                entry.Size = new FileInfo(absolute).Length;
                entries.Add(entry);
            }

            foreach (string subDir in Directory.GetDirectories(dir))
            {
                string name = new DirectoryInfo(subDir).Name;
                Walk(subDir, relPrefix + "/" + name, entries);
            }
        }

        // ВАЖНО: это чисто лексическая нормализация (Path.GetFullPath убирает "..",
        // ".", относительность), а не разыменование symlink/junction — аналога
        // Swift-стороннего resolvingSymlinksInPath() в .NET Framework 4.0 без P/Invoke
        // нет. Согласованность размера и байтов держится не на этом: она держится на
        // том, что ОДНА И ТА ЖЕ возвращённая строка кладётся и в Entry.AbsolutePath
        // (источник для File.ReadAllBytes в Blob()), и передаётся в FileInfo(...).Length
        // для BlobRef.Size — расхождение между объявленным размером и прочитанными
        // байтами структурно исключено именно поэтому, а не потому что путь "разрешён".
        static string CanonicalizeForBothReads(string path)
        {
            return Path.GetFullPath(path);
        }
    }
}
